[while_loop] auto turn int inputs into unbacked symint

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #141393
* __->__ #140106
* #142031
* #142162

Previously, only tensor carries are allowed. After this PR, users can now pass in an integer as carry directly:
```python
def cond_fn(idx, x):
  return idx < x.shape[0]

# This implements a map
def body_fn(idx, x):
  out = x.clone()
  out[idx] = x[idx] + 1
  return idx + 1, out 

while_loop(cond_fn, body_fn (0, torch.randn(3, 4)))
```
**Implementation-wise:**
Every step of compilation (dynamo, aot, inductor) does similar things: 1. we unspecialize constant int carries by creating unbacked symints. (can automatically add torch._check based on cond_fn's return) 2. trace/lower the subgraph with the unbacked symints we created. 3. we allocate new unbakced symints as outputs and finish lowering of the hop.



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

ydwu4 committed:
[while_loop] auto turn int inputs into unbacked symint

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] auto turn int inputs into unbacked symint"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -3716,13 +3742,9 @@ def test_while_loop_nested2_traced(self):
     while_loop_cond_graph_0 = self.while_loop_cond_graph_0
     while_loop_body_graph_0 = self.while_loop_body_graph_0
-    while_loop = torch.ops.higher_order.while_loop(while_loop_cond_graph_0, while_loop_body_graph_0, (arg0_1, arg1_1, arg2_1, arg3_1), (sym_size_int, sym_size_int_1, sym_size_int_2, sym_size_int_3));  while_loop_cond_graph_0 = while_loop_body_graph_0 = arg0_1 = arg1_1 = arg2_1 = arg3_1 = sym_size_int = sym_size_int_1 = sym_size_int_2 = sym_size_int_3 = None
+    while_loop = torch.ops.higher_order.while_loop(while_loop_cond_graph_0, while_loop_body_graph_0, (arg0_1, arg1_1, arg2_1, arg3_1), ());  while_loop_cond_graph_0 = while_loop_body_graph_0 = arg0_1 = arg1_1 = arg2_1 = arg3_1 = None
```
Because we no longer test with symbolic, no symints used are lifted.

zou3519 commented on torch/_dynamo/source.py:
```diff
@@ -616,6 +616,15 @@ def guard_source(self):
         return GuardSource.GLOBAL
 
 
+@dataclasses.dataclass(frozen=True)
+class DummyHopPlaceholderSource(Source):
```
what do you we need this for?

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1029,64 +1109,70 @@ def call_function(
+                (cond_r, cond_treespec),
+                cond_graph,
+                cond_lifted_freevars,
+            ) = speculate_subgraph(
```
The way I would have implemented this is:
- for each int in operands_seq, create new unbacked symints
- call speculate subgraph on the the new operands_seq.

The thing that confuses me are:
- why does this API need to be a context manager?
- why do we need to create dummy placeholders for the unbacked symint?

ydwu4 replied:
> why do we need to create dummy placeholders for the unbacked symint?

This is to workaround the invariant that each SymNode variable need to have a proxy. I created a placeholder for this purpose. I guess I could create a call function proxy (e.g. create_unbacked_symint) and remove it after finish speculating the subgraph but it's essentially the same thing.

> why does this API need to be a context manager?
we want to delete the dummy placeholder after speculation.

zou3519 replied:
flexattention also has a problem where it just creates random dummy proxies. Can we have some sort of generic helper function for temporarily creating proxies and then deleting them?

Since that seems to be where we're going in general with HOP tracing ("temporarily create new proxies, trace with them, and then throw them out") that should probably be the main API rather than adding new placeholders

ydwu4 replied:
yeah, that's true. 

Let me think about this. One approach is to do this automatically in speculate_subgraph but we need to design the interface to allow customizations (e.g. we don't always want to cast int to unbacked symint)

zou3519 replied:
we probably also do this in map, right? We need to slice the inputs so that we can speculate_subgraph on the sliced inputs

ydwu4 replied:
Yeah, scan/associative_scan also works on a slice of inputs. basically, any hops that's not wrap-like, might potentially need to create temp variables.

So I feel concretely the speculate_subgraph should take **exmaple_values**  (e.g. fak tensors, symints) instead of variable trackers (kind of like the builder.wrap). Because what we do right now in validate_args_and_maybe_create_graph_inputs is just extracting "exmaple_value" and create graph_input using the example_value.

ydwu4 replied:
oh, we can't we because we need the variable tracker to call " f.call_function(tx, args, sub_kwargs)"

ydwu4 committed:
Update on "[while_loop] auto turn int inputs into unbacked symint"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] auto turn int inputs into unbacked symint"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] auto turn int inputs into unbacked symint"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] auto turn int inputs into unbacked symint"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] auto turn int inputs into unbacked symint"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] auto turn int inputs into unbacked symint"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] auto turn int inputs into unbacked symint"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] auto turn int inputs into unbacked symint"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] auto turn int inputs into unbacked symint"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] auto turn int inputs into unbacked symint"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] auto turn int inputs into unbacked symint"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] auto turn int inputs into unbacked symint"


Previously, only tensor carries are allowed. After this PR, users can now pass in an integer as carry directly:
```python
def cond_fn(idx, x):
  return idx < x.shape[0]

# This implements a map
def body_fn(idx, x):
  out = x.clone()
  out[idx] = x[idx] + 1
  return idx + 1, out 

while_loop(cond_fn, body_fn (0, torch.randn(3, 4)))
```
**Implementation-wise:**
Every step of compilation (dynamo, aot, inductor) does similar things: 1. we unspecialize constant int carries by creating unbacked symints. (can automatically add torch._check based on cond_fn's return) 2. trace/lower the subgraph with the unbacked symints we created. 3. we allocate new unbakced symints as outputs and finish lowering of the hop.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] auto turn int inputs into unbacked symint"


Previously, only tensor carries are allowed. After this PR, users can now pass in an integer as carry directly:
```python
def cond_fn(idx, x):
  return idx < x.shape[0]

# This implements a map
def body_fn(idx, x):
  out = x.clone()
  out[idx] = x[idx] + 1
  return idx + 1, out 

while_loop(cond_fn, body_fn (0, torch.randn(3, 4)))
```
**Implementation-wise:**
Every step of compilation (dynamo, aot, inductor) does similar things: 1. we unspecialize constant int carries by creating unbacked symints. (can automatically add torch._check based on cond_fn's return) 2. trace/lower the subgraph with the unbacked symints we created. 3. we allocate new unbakced symints as outputs and finish lowering of the hop.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] auto turn int inputs into unbacked symint"


Previously, only tensor carries are allowed. After this PR, users can now pass in an integer as carry directly:
```python
def cond_fn(idx, x):
  return idx < x.shape[0]

# This implements a map
def body_fn(idx, x):
  out = x.clone()
  out[idx] = x[idx] + 1
  return idx + 1, out 

while_loop(cond_fn, body_fn (0, torch.randn(3, 4)))
```
**Implementation-wise:**
Every step of compilation (dynamo, aot, inductor) does similar things: 1. we unspecialize constant int carries by creating unbacked symints. (can automatically add torch._check based on cond_fn's return) 2. trace/lower the subgraph with the unbacked symints we created. 3. we allocate new unbakced symints as outputs and finish lowering of the hop.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] auto turn int inputs into unbacked symint"


Previously, only tensor carries are allowed. After this PR, users can now pass in an integer as carry directly:
```python
def cond_fn(idx, x):
  return idx < x.shape[0]

# This implements a map
def body_fn(idx, x):
  out = x.clone()
  out[idx] = x[idx] + 1
  return idx + 1, out 

while_loop(cond_fn, body_fn (0, torch.randn(3, 4)))
```
**Implementation-wise:**
Every step of compilation (dynamo, aot, inductor) does similar things: 1. we unspecialize constant int carries by creating unbacked symints. (can automatically add torch._check based on cond_fn's return) 2. trace/lower the subgraph with the unbacked symints we created. 3. we allocate new unbakced symints as outputs and finish lowering of the hop.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] auto turn int inputs into unbacked symint"


Previously, only tensor carries are allowed. After this PR, users can now pass in an integer as carry directly:
```python
def cond_fn(idx, x):
  return idx < x.shape[0]

# This implements a map
def body_fn(idx, x):
  out = x.clone()
  out[idx] = x[idx] + 1
  return idx + 1, out 

while_loop(cond_fn, body_fn (0, torch.randn(3, 4)))
```
**Implementation-wise:**
Every step of compilation (dynamo, aot, inductor) does similar things: 1. we unspecialize constant int carries by creating unbacked symints. (can automatically add torch._check based on cond_fn's return) 2. trace/lower the subgraph with the unbacked symints we created. 3. we allocate new unbakced symints as outputs and finish lowering of the hop.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]
